support mount multiple volumes

l646505418 opened the pull request:
## Description

Add voluimeClaimTemplates to Storage.Spec to support mount multiple volumes

#5578 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [`x` ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

nicolastakashi commented:
Amazing work here @l646505418!
I'm just wondering if we should have some deprecation process on the `VolumeClaimTemplate`

l646505418 commented:
> 

As @xiu commented before, `Since Prometheus is a v1 CRD, we can't remove a field without bumping the version (e.g. v2). In this case, we'd have to add a new field but not remove the previous one for compatibility purposes in order to stick to v1.`

xiu commented:
👍 we can update the comment to direct users to use the new field without removing the old one though.

l646505418 commented:
> 👍 we can update the comment to direct users to use the new field without removing the old one though.

Yes, but as I mentioned before, I met some diffculties to run `make generate`, it is hard for me to generate CRDs and other files(eg. zz_generated)

simonpasquier commented:
For the record, the k8s developer documentation has some recommendations about making a singular field plural:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#making-a-singular-field-plural

I'm not sure how it works in practice for CRDs (vs. k8s native resources).

xiu commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -838,6 +838,8 @@ type StorageSpec struct {
 	// The easiest way to use a volume that cannot be automatically provisioned
```
This comment should be led by something along the lines of:
```*DEPRECATED*: Please use `VolumeClaimTemplates` which allows for multiple VolumeClaim to be used.```

xiu commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -838,6 +838,8 @@ type StorageSpec struct {
 	// The easiest way to use a volume that cannot be automatically provisioned
 	// is to use a label selector alongside manually created PersistentVolumes.
 	VolumeClaimTemplate EmbeddedPersistentVolumeClaim `json:"volumeClaimTemplate,omitempty"`
+
```
Add a comment explaining what this field does, probably same as the previous one (without the deprecation notice) but plural.

xiu commented on pkg/prometheus/agent/statefulset.go:
```diff
@@ -137,6 +137,22 @@ func makeStatefulSet(
 				Ephemeral: ephemeral,
 			},
 		})
+	} else if storageSpec.VolumeClaimTemplates != nil {
+		for _, specVcTemplate := range storageSpec.VolumeClaimTemplates {
```
I didn't test it but we might be able to simplify this part by removing the else block on L140 and supporting both in this one. (same comment for the server)

```suggestion
	} else {
	    if storageSpec.VolumeClaimTemplate != nil {
	        storageSpec.VolumeClaimTemplates = append(storageSpec.VolumeClaimTemplates, storageSpec.VolumeClaimTemplate)
	    }

	    for _, specVcTemplate := range storageSpec.VolumeClaimTemplates {
```

What do you think?

xiu reviewed (commented):
Thanks for your work on this, @l646505418! Had another look at this one, I added comments where I think we should update that comment.

> Yes, but as I mentioned before, I met some diffculties to run make generate, it is hard for me to generate CRDs and other files(eg. zz_generated)

We'll help with that if we can't find another solution no worries. 

I see we have 3 unit tests failing at the moment:
* `TestStatefulSetVolumeInitial`
* `TestThanosObjectStorage`
* `TestThanosObjectStorageFile`

Once those tests are fixed and comments addressed, we'll make sure to generate assets to see the pipeline go green before merging 👍

l646505418 commented:
> Thanks for your work on this, @l646505418! Had another look at this one, I added comments where I think we should update that comment.
> 
> > Yes, but as I mentioned before, I met some diffculties to run make generate, it is hard for me to generate CRDs and other files(eg. zz_generated)
> 
> We'll help with that if we can't find another solution no worries.
> 
> I see we have 3 unit tests failing at the moment:
> 
> * `TestStatefulSetVolumeInitial`
> * `TestThanosObjectStorage`
> * `TestThanosObjectStorageFile`
> 
> Once those tests are fixed and comments addressed, we'll make sure to generate assets to see the pipeline go green before merging 👍

Hi，I have updated codes, can you guys review again?

ArthurSens commented on pkg/prometheus/server/statefulset.go:
```diff
@@ -201,18 +202,25 @@ func makeStatefulSet(
-		pvcTemplate := operator.MakeVolumeClaimTemplate(storageSpec.VolumeClaimTemplate)
-		if pvcTemplate.Name == "" {
-			pvcTemplate.Name = prompkg.VolumeName(p)
+		if reflect.DeepEqual(storageSpec.VolumeClaimTemplate, monitoringv1.EmbeddedPersistentVolumeClaim{}) {
+			storageSpec.VolumeClaimTemplates = append([]monitoringv1.EmbeddedPersistentVolumeClaim{storageSpec.VolumeClaimTemplate}, storageSpec.VolumeClaimTemplates...)
```
As far as I could understand, you're comparing if VolumeClaimTemplate is equal com an empty object? And only if it is indeed empty, then you're appending it to array of PVClaims?

Seen in [godocs](https://pkg.go.dev/reflect#DeepEqual) of DeeplyEqual:
`Struct values are deeply equal if their corresponding fields, both exported and unexported, are deeply equal.`

```suggestion
		if !reflect.DeepEqual(storageSpec.VolumeClaimTemplate, monitoringv1.EmbeddedPersistentVolumeClaim{}) {
			storageSpec.VolumeClaimTemplates = append([]monitoringv1.EmbeddedPersistentVolumeClaim{storageSpec.VolumeClaimTemplate}, storageSpec.VolumeClaimTemplates...)
```
Looks like we're only missing the negation operator here :P

ArthurSens reviewed (commented):
Hey @l646505418 , thanks for working on this. It's looking like we're in a good direction, just need to fix a few things that were caught in out tests.

Ideally, you'd be able to run those tests yourself by running `make test-unit` or `make test-e2e` and there's no need to depend on CI to catch them during development. Most editors also provide a good debugging experience when working with go projects :)

Let us know if you're also having problems running tests!

l646505418 commented:
> 

Oh， your change is right, it is my fault😂
I indeed can run unit test in my local, but I have mentioned before I have some troubles generating CRDs , I am not sure if that will have big impacts.

xiu commented:
@l646505418 are you able to run `make test-unit` in your branch? We still see 3 failing tests, they shouldn't be related to `make generate`.

l646505418 commented:
> @l646505418 are you able to run `make test-unit` in your branch? We still see 3 failing tests, they shouldn't be related to `make generate`.

Hi ,I modified codes and all unit tests are passed, can you review?

ArthurSens commented on pkg/prometheus/agent/statefulset.go:
```diff
@@ -138,18 +139,24 @@ func makeStatefulSet(
-			pvcTemplate.Spec.AccessModes = []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}
-		} else {
-			pvcTemplate.Spec.AccessModes = storageSpec.VolumeClaimTemplate.Spec.AccessModes
+		for _, specVcTemplate := range storageSpec.VolumeClaimTemplates {
```
```suggestion
		for _, vcTemplate := range storageSpec.VolumeClaimTemplates {
```

this is just a volumeClaimTemplate, feels confusing to have spec on the variable name 🤔

ArthurSens commented on pkg/prometheus/server/statefulset_test.go:
```diff
@@ -228,6 +228,69 @@ func TestStatefulSetPVC(t *testing.T) {
+	require.NoError(t, err)
+
+	if !reflect.DeepEqual(2, len(sset.Spec.VolumeClaimTemplates)) {
+		t.Fatal("Error adding PVCs Spec to StatefulSetSpec")
```
```suggestion
		t.Fatalf("Expected %d PVCs, got %d", 2, len(sset.Spec.VolumeClaimTemplates))
```

ArthurSens commented on pkg/prometheus/server/statefulset_test.go:
```diff
@@ -228,6 +228,69 @@ func TestStatefulSetPVC(t *testing.T) {
+
+	if !reflect.DeepEqual(2, len(sset.Spec.VolumeClaimTemplates)) {
+		t.Fatal("Error adding PVCs Spec to StatefulSetSpec")
+	} else {
+		ssetPvc := sset.Spec.VolumeClaimTemplates[0]
+		if !reflect.DeepEqual(*pvc.Spec.StorageClassName, *ssetPvc.Spec.StorageClassName) {
+			t.Fatal("Error adding PVCs Spec to StatefulSetSpec")
+		}
+
+		ssetPvcSecond := sset.Spec.VolumeClaimTemplates[1]
+		if !reflect.DeepEqual(*pvcSecond.Spec.StorageClassName, *ssetPvcSecond.Spec.StorageClassName) {
+			t.Fatal("Error adding PVCs Spec to StatefulSetSpec")
+		}
+	}
```
We can make the code more readable by reducing the indentation level here. (removing unnecessary else statement)

```suggestion
	} 
	ssetPvc := sset.Spec.VolumeClaimTemplates[0]
	if !reflect.DeepEqual(*pvc.Spec.StorageClassName, *ssetPvc.Spec.StorageClassName) {
		t.Fatal("Error adding PVCs Spec to StatefulSetSpec")
	}

	ssetPvcSecond := sset.Spec.VolumeClaimTemplates[1]
	if !reflect.DeepEqual(*pvcSecond.Spec.StorageClassName, *ssetPvcSecond.Spec.StorageClassName) {
		t.Fatal("Error adding PVCs Spec to StatefulSetSpec")
	}
```

ArthurSens replied:
Also, why are we comparing only StorageClassName? Shouldn't we compare the whole PVC?

ArthurSens commented on pkg/prometheus/statefulset.go:
```diff
@@ -366,14 +369,29 @@ func BuildCommonVolumes(p monitoringv1.PrometheusInterface, tlsAssetSecrets []st
+	volNames := []string{}
 	if cpf.Storage != nil {
-		if cpf.Storage.VolumeClaimTemplate.Name != "" {
+		volName := VolumeName(p)
```
```suggestion
```
This value isn't used since we override it on line 378.

If you need to initialize the variable `volName`, you can declare it at the beginning of this function:

```go
func VolumeClaimName(p monitoringv1.PrometheusInterface, cpf monitoringv1.CommonPrometheusFields) []string {
	volNames := []string{}
	var volName string
.
.
.
```

ArthurSens commented on pkg/prometheus/statefulset.go:
```diff
@@ -366,14 +369,29 @@ func BuildCommonVolumes(p monitoringv1.PrometheusInterface, tlsAssetSecrets []st
+				}
+				volNames = append(volNames, volName)
+			}
+		} else if cpf.Storage.VolumeClaimTemplate.Name != "" {
```
The only way we get here is if `cpf.Storage.VolumeClaimTemplates` is nil, right? If that's the case, trying to access `cpf.Storage.VolumeClaimTemplate.Name` will throw a nil pointer exception. 

We should do nothing if VolumeClaimTemplates is nil

l646505418 replied:
So how about when user only define `Storage.VolumeClaimTemplate`?

ArthurSens replied:
Ah sorry, I misread the condition. I thought this line was comparing VolumeClaimTemplate<ins>s</ins>.Name 😅 

You're correct

ArthurSens commented on pkg/prometheus/statefulset.go:
```diff
@@ -366,14 +369,29 @@ func BuildCommonVolumes(p monitoringv1.PrometheusInterface, tlsAssetSecrets []st
 		}
+	} else {
+		volName := VolumeName(p)
+		volNames = append(volNames, volName)
```
Same thing here. If `Storage` is nil, we don't want to append any volume.

ArthurSens reviewed (approved):
LGTM, thanks!

ArthurSens commented:
Good find!

I'm reverting the PR and let's discuss the new approach
